### ISSUE TITLE
Prevent user from entering classroom directly

### ISSUE THREAD

**Issue by mssiegel** [mssiegel/frempco-client#86]
Currently a user can directly visit Teachers Page: www.frempco.com/teacher/classroom/apple
Currently, a user can currently directly visit Students Page: www.frempco.com/student/classroom/apple

We want all users to enter their classroom only from the homepage.
- Teachers page: Prevents people from being able to enter teachers classroom without a password
- Students page: Prevents students from visiting the link and thinking they're logged in as a student when they're not

See converation in: #80 
Perhaps we can use browser session storage to see if the student came from the frempco homepage. If not, we can redirect them to the homepage. See:  https://www.grouparoo.com/blog/getting-previous-path-nextjs

**Comment by zachlankton** [comment]
Here is my proposal that doesn't need page history tracking.  In our `_app.tsx` we could check what page the app starts on and if it is not `/` then we force the user to the homepage.  I have tested the following code and it works quite well for this (I can do a PR if you like this idea)

```javascript
export default function MyApp(props: MyAppProps) {
  const router = useRouter();
  const [userIsAuthorized, setUserIsAuth] = React.useState(false);
  React.useEffect(() => {
    if (router.pathname != '/') {
      window.location.assign('/');
    } else {
      setUserIsAuth(true);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

...
}
```

Disabling the exhaustive deps lint so we can pass in an empty array will ensure this code only runs one time on component mount (In this case: our entire app)

Using `window.location.assign` vs `router.push` ensures that the `setUserIsAuth` code will run when the app starts on the homepage which can then be used to actually display the app.

The extra step of using the `userIsAuthorized` state helps prevent a brief flash of the teachers page or student page before getting forced to the home page.

```javascript
 if (userIsAuthorized)
    return (
      <CacheProvider value={emotionCache}>
        <Head>
          <title>Create Next App</title>
          <meta name='viewport' content='initial-scale=1, width=device-width' />
        </Head>
        <ThemeProvider theme={theme}>
          {/* CssBaseline kickstart an elegant, consistent, and simple baseline to build upon. */}
          <CssBaseline />
          <SocketProvider>
            <UserProvider>
              <Component {...pageProps} />
            </UserProvider>
          </SocketProvider>
        </ThemeProvider>
      </CacheProvider>
    );

  return <></>
```

**Comment by mssiegel** [comment]
This is rather brilliant!! Right now we only have TeachersPage and StudentsPage. But eventually we'll be adding more pages, perhaps an "About us" page that we're okay with people going directly to.  Will this fix still work in that situation?

**Comment by zachlankton** [comment]
Absolutely, just a tiny tweak... we could have a whitelist array of allowed pages and check like so...

```javascript
const whitelist = [
  '/',
  '/about',
  '/etc'
]

...

if ( !whitelist.includes(router.pathname) ) {

...

```

**Comment by mssiegel** [comment]
Oh this is brilliant!! I'd prefer we use a `denylist`, because that list will be shorter in the long run. Is that possible?
If not we'd use the whitelist. 
Feel free to go ahead and make this PR!

**Comment by zachlankton** [comment]
Yep denylist works too! I'll get this PR'd right away!

**Comment by mssiegel** [comment]
I approved the PR. This comment should not block merging of PR as this issue needs to get solved before Ariella uses it again on Monday. I'm only writing this for us to think about for later, perhaps when we meet on Monday. 

I'm thinking this is part of a broader idea that some routes should be restricted to logged in users only. Most codebases would store authentication via server side, but we need something to do so client side that could work with our UserContext. 

When a student or teacher logins, we'd store that in our UserContext. I found this guide to implementing [**client side** authentication redirects in nextjs](https://azeezatraheem.medium.com/implementing-authentication-redirects-in-next-js-c15907ec82b7).
